### ISSUE TITLE
Difference between GB's and classic editor's interpretation of TinyMCE's style_formats option

### ISSUE THREAD

**Issue by pbiron** [WordPress/gutenberg#8670]
**Describe the bug**
There is a difference in how GB v3.4 interprets the value of TinyMCE's `style_formats` option for classic blocks and how the classic editor does (with or without the `Classic Editor` plugin).

**To Reproduce**
1. Hook into core's `tiny_mce_before_init` and return a value for `style_formats` conforming to the examples in the [TinyMCE docs](https://www.tiny.cloud/docs/configure/content-formatting/#style_formats)
2. Edit/create a post in the classic editor and see that the `Formats` dropdown has the styles specified in the return from `tiny_mce_before_init`
3. Edit/create a post in GB and insert a `classic` block
4. See that the items in the `Formats` dropdown are different than in the classic editor

**Expected behavior**
The options in the `Formats` dropdown should be the same in the classic editor and GB

**Screenshots**
Setting in [Advanced TinyMCE Configuation](https://wordpress.org/plugins/advanced-tinymce-configuration/) I've used on 1 particular site for years, this corresponds to step 1 above

![tinymce_advanced_config](https://user-images.githubusercontent.com/3824560/43791017-90e33258-9a31-11e8-9a9a-02c02ff9d015.png)

TinyMCE in core for that particular site, this corresponds to step 2 above

![tinymce_core](https://user-images.githubusercontent.com/3824560/43791293-708a0fa8-9a32-11e8-8287-1b1cc080142b.png)

GB for that particular site that shows the difference, , this corresponds to step 3-4 above

![gb_difference](https://user-images.githubusercontent.com/3824560/43791265-5da5a294-9a32-11e8-8111-d586f56a487b.png)

Alternate setting in Advanced TinyMCE Configuation that works in both classic and GB (because it is `json_decode()`able

![tinymce_advanced_config_alternate](https://user-images.githubusercontent.com/3824560/43791349-a4c159fc-9a32-11e8-93bc-a23a6beb0715.png)

TinyMCE in core with the alternate `style_formats` setting

![tinymce_core_with_alternate](https://user-images.githubusercontent.com/3824560/43791382-bf8fe7c6-9a32-11e8-93c3-720c0002fd70.png)

GB with the alternate `style_formats` setting.  .  Note that although the correct "styles" are shown in the dropdown, they are **not** styled as they are in TinyMCE.  At this point, I'm not too concerned about that point, but I can certainly add a separate issue if needed.

![gb_alternate](https://user-images.githubusercontent.com/3824560/43791398-cdca18d4-9a32-11e8-8d51-3f9f9e18ed3a.png)

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro
 - Browser: FF 57, Chrome 68

**Additional context**
The problem is that `gutenberg_register_scripts_and_styles()` does `json_decode( $tinymce_settings['style_formats'] )` and TinyMCE does **not** require that the value of `style_formats` be json "decodable".  In fact, if you `json_decode()` any of the `style_formats` examples in the TinyMCE docs they all return `null`, which accounts for the difference in behavior between classic and GB.

Luckily, TinyMCE accepts values that **are** json decodable.  So at a minimum this difference should be  **documented**.  But, of course, it would be better if a solution could be found such that the same `style_formats` value would work for classic editor and GB.

**Comment by danielbachhuber** [comment]
Related https://github.com/danielbachhuber/gutenberg-migration-guide/issues/10

**Comment by pbiron** [comment]
I haven't checked whether this problem still exists in GB 3.9.0, I'll try to do so in the next few days and report back.

**Comment by pbiron** [comment]
this problem still exists in both GB 3.9 & 4.0.

this was discussed during the [bug scrub today](https://wordpress.slack.com/archives/C02RQBWTW/p1539800830000100).  @mtias suggested that it was a "documentation task".

I mentioned when I opened the issue that

> Luckily, TinyMCE accepts values that are json decodable. So at a minimum this difference should be documented. But, of course, it would be better if a solution could be found such that the same style_formats value would work for classic editor and GB.

Also, as I said originally, **IF** you use a json decodable value for `style_formats` it works but the menu items are **not** styled in GB like they are in Classic.  Which it ultimately would be great to solve (but that can wait to WP 5.1 if necessary).

**Comment by azaozz** [comment]
As this is already merged in core opened https://core.trac.wordpress.org/ticket/45117 as a follow up.

tl;dr: we need to drop the `json_decode()` and use the same way of "translating" the TinyMCE settings from PHP to JS as in class-wp-editor.php.

**Comment by danielbachhuber** [comment]
Being fixed with #10966

@pbiron Can you test #10968 and verify?

**Comment by pbiron** [comment]
> Being fixed with #10966
> 
> @pbiron Can you test #10968 and verify?

I can confirm that with that patch in #10968 applied to GB 4.1.1 that _style_format_ values that conform to the [TinyMCE docs](https://www.tiny.cloud/docs/configure/content-formatting/#style_formats) are now displayed in the _Formats_ dropdown.  Good work @azaozz!

**Comment by pbiron** [comment]
While #10968 allows the correct options in the _Formats_ dropdown to appear, they are still **not** styled according to their respective definitions, as mentioned in my original description of this issue:

> GB with the alternate style_formats setting. . Note that although the correct "styles" are shown in the dropdown, they are not styled as they are in TinyMCE.

I think as soon as #10968 is merged this issue can be closed and I will open another issue for the styling of the options.

**Comment by youknowriad** [comment]
closed by #10968
